research/deep_speech/data/download.py

```diff
@@ -0,0 +1,167 @@
+#  Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+# ==============================================================================
+"""Download and preprocess LibriSpeech dataset for DeepSpeech model."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import codecs
+import csv
+import fnmatch
+import os
+import sys
+import tarfile
+import tempfile
+import unicodedata
+
+from absl import flags as absl_flags
+from six.moves import urllib
+from sox import Transformer
+import tensorflow as tf
+
+FLAGS = absl_flags.FLAGS
+
+absl_flags.DEFINE_string(
+    "data_dir", "/tmp/librispeech_data",
+    "Directory to download data and extract the tarball")
+
+LIBRI_SPEECH_URLS = {
+    "train-clean-100":
+        "http://www.openslr.org/resources/12/train-clean-100.tar.gz",
+    "train-clean-360":
+        "http://www.openslr.org/resources/12/train-clean-360.tar.gz",
+    "train-other-500":
+        "http://www.openslr.org/resources/12/train-other-500.tar.gz",
+    "dev-clean":
+        "http://www.openslr.org/resources/12/dev-clean.tar.gz",
+    "dev-other":
+        "http://www.openslr.org/resources/12/dev-other.tar.gz",
+    "test-clean":
+        "http://www.openslr.org/resources/12/test-clean.tar.gz",
+    "test-other":
+        "http://www.openslr.org/resources/12/test-other.tar.gz"
+}
+
+
+def download_and_extract(directory, url):
+  """Download and extract the given split of dataset.
+  Args:
+    directory: the directory where to extract the tarball.
+    url: the url to download the data file.
+  """
+
+  if not tf.gfile.Exists(directory):
+    tf.gfile.MakeDirs(directory)
+  _, tar_filepath = tempfile.mkstemp(suffix=".tar.gz")
```

Comment:
I think the try-final block should start here, otherwise it does not ensure the cleanup of the temp file.

Reply:
Thanks. Done.